iree/hal/vulkan/vma_allocator.cc

```diff
@@ -56,100 +82,135 @@ StatusOr<std::unique_ptr<VmaAllocator>> VmaAllocator::Create(
-    Allocator* source_allocator, iree_hal_memory_type_t memory_type,
-    iree_hal_buffer_usage_t buffer_usage,
-    iree_hal_buffer_usage_t intended_usage) const {
-  // TODO(benvanik): ensure there is a memory type that can satisfy the request.
-  return source_allocator == this;
+static iree_allocator_t iree_hal_vulkan_vma_allocator_host_allocator(
+    const iree_hal_allocator_t* base_allocator) {
+  iree_hal_vulkan_vma_allocator_t* allocator =
+      (iree_hal_vulkan_vma_allocator_t*)base_allocator;
+  return allocator->host_allocator;
 }
 
-bool VmaAllocator::CanAllocate(iree_hal_memory_type_t memory_type,
-                               iree_hal_buffer_usage_t buffer_usage,
-                               size_t allocation_size) const {
-  // TODO(benvnik): ensure there is a memory type that can satisfy the request.
-  return true;
+static iree_hal_buffer_compatibility_t
+iree_hal_vulkan_vma_allocator_query_buffer_compatibility(
+    iree_hal_allocator_t* base_allocator, iree_hal_memory_type_t memory_type,
+    iree_hal_buffer_usage_t allowed_usage,
+    iree_hal_buffer_usage_t intended_usage,
+    iree_device_size_t allocation_size) {
+  iree_hal_vulkan_vma_allocator_t* allocator =
+      iree_hal_vulkan_vma_allocator_cast(base_allocator);
+
+  // TODO(benvanik): check to ensure the allocator can serve the memory type.
+
+  // Disallow usage not permitted by the buffer itself. Since we then use this
+  // to determine compatibility below we'll naturally set the right compat flags
+  // based on what's both allowed and intended.
+  intended_usage &= allowed_usage;
+
+  // All buffers can be allocated on the heap.
+  iree_hal_buffer_compatibility_t compatibility =
+      IREE_HAL_BUFFER_COMPATIBILITY_ALLOCATABLE;
+
+  // Buffers can only be used on the queue if they are device visible.
+  if (iree_all_bits_set(memory_type, IREE_HAL_MEMORY_TYPE_DEVICE_VISIBLE)) {
+    if (iree_all_bits_set(intended_usage, IREE_HAL_BUFFER_USAGE_TRANSFER)) {
+      compatibility |= IREE_HAL_BUFFER_COMPATIBILITY_QUEUE_TRANSFER;
+    }
+    if (iree_all_bits_set(intended_usage, IREE_HAL_BUFFER_USAGE_DISPATCH)) {
+      compatibility |= IREE_HAL_BUFFER_COMPATIBILITY_QUEUE_DISPATCH;
+    }
+  }
+
+  return compatibility;
 }
 
-Status VmaAllocator::MakeCompatible(
+static iree_status_t iree_hal_vulkan_vma_allocator_make_compatible(
     iree_hal_memory_type_t* memory_type,
-    iree_hal_buffer_usage_t* buffer_usage) const {
-  // TODO(benvanik): mutate to match supported memory types.
-  return OkStatus();
+    iree_hal_memory_access_t* allowed_access,
+    iree_hal_buffer_usage_t* allowed_usage) {
+  // TODO(benvanik): remove this entirely!
+  // Host currently uses mapping to copy buffers, which is done a lot.
+  // We could probably remove this mutation by preventing copies in those cases
+  // or issuing small copy command buffers.
+  *allowed_usage |=
+      IREE_HAL_MEMORY_TYPE_HOST_VISIBLE | IREE_HAL_BUFFER_USAGE_MAPPING;
```

Comment:
This is marking all VMA-allocated buffers as host visible?

Reply:
yep ;( this is required right now because of the fill/copy/etc methods mapping the memory and our silly vm_utils test infra directly touching buffers instead of using staging buffers. If we can kill the fill/copy/etc methods (making them always go through command buffers) we can remove this.